feat(network-creation): macvlan driver for swarm mode (#2082)

xAt0mZ opened the pull request:
Adding the MACVLAN driver for network creation in Portainer.

Closes #2082 .

ncresswell commented:
OK, so the UX for this is quite hard to understand... we need to find a way to dramatically simplify. I have used MACLAN a number of times, and even i got lost trying to follow the UI.

Did you take a look at how UCP addresses MACVLAN definition ? Theirs seems simpler.

ncresswell commented:
if you select macvlan, and select "creation" then you should not be prompted for Network Configuration again (subnet etc)

ncresswell commented:
creating a MACVLAN config does not force you to enter a value for parent network card, yet this is a requirement for MACVLAN to work. that needs to be a mandatory field

ncresswell commented:
sorry, one more. dont use the term "aux address" as its meaningless to most people, use "exclude IPs" or something, which reflects that the IP entered in that field is excluded from the IP Pool

xAt0mZ commented:
I'm going to rework the entire view, as we should not hide "above" elements when clicking a field somewhere in a page -> with the actual layout we can't hide IPAM (network) config if a user clicks the `MACVLAN > Creation` section (we can technically, but we don't want to as it's not UX/UI friendly).

UCP has the same layout than we have but in a different order (links below).
For configuration : <details>
![2018-08-01-143845_957x601_scrot](https://user-images.githubusercontent.com/7642331/43711903-94d8cf5c-9974-11e8-8337-08e5e4215b88.png)
![2018-08-01-143920_376x496_scrot](https://user-images.githubusercontent.com/7642331/43711902-94c0ad0a-9974-11e8-8a32-9193227c6216.png)
![2018-08-01-143953_704x701_scrot](https://user-images.githubusercontent.com/7642331/43711901-94aa66c6-9974-11e8-965b-eb3030317a7f.png)</details>

For deployment : <details>
![2018-08-01-144009_416x480_scrot](https://user-images.githubusercontent.com/7642331/43711945-b8b5fb2a-9974-11e8-80d1-e08d68e0f342.png)
![2018-08-01-143920_376x496_scrot](https://user-images.githubusercontent.com/7642331/43711947-b8e66b84-9974-11e8-9da7-27ece1c5706f.png)
![2018-08-01-143953_704x701_scrot](https://user-images.githubusercontent.com/7642331/43711946-b8cf6696-9974-11e8-80fb-c2d1c7830442.png)</details>


They don't have Network Card selection, and it doesn't seem to be mandatory on the examples I found about MACVLAN usage ; still I can require it.

xAt0mZ commented:
@ncresswell @deviantony UI reworked, waiting for you feedback !

deviantony commented:
@xAt0mZ awesome, I'll rebuild the test image.

deviantony commented:
@ncresswell can you validate this one functionally ? You can use `portainer/portainer:pr2122` to try it.

ncresswell commented:
I cannot get this to work.. every time I try I get “gateway address already in use” error

From the CLI, this is how I create my MACVLAN config:
docker network create --config-only --subnet 192.168.1.0/24 --gateway 192.168.1.1 --ip-range 192.168.1.120/32 -o parent=ens160 macvlan_config

And then this is how I create the MACVLAN Network
docker network create -d macvlan --scope swarm --attachable --config-from macvlan_config macvlan_network

which works.

But when I try via the new UI,

I set the parent to ens160, I set the subnet to be 192.168.1.0/24, the gateway to be 192.168.1.1, the range to be 192.168.1.120/32.

The creation succeeds, but when I create a service and attach a container to that network, the tasks are always rejected with “gateway address already in use”.

I tried adding in the gateway to the AUX address, same error.

What am I doing wrong ?



From: Anthony Lapenna <notifications@github.com>
Sent: Tuesday, 7 August 2018 1:17 a.m.
To: portainer/portainer <portainer@noreply.github.com>
Cc: Neil Cresswell <neil@cresswell.net.nz>; Mention <mention@noreply.github.com>
Subject: Re: [portainer/portainer] feat(network-creation): macvlan driver for swarm mode (#2082) (#2122)


@ncresswell<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ncresswell&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=0fx0h4vB56iTLpw2McH1ZD6TqG_QGpbggVOB-PfMJpM&m=uHSnTkysep872hMF0-ElTCSUQZTqyC2TVr-PsxPC83g&s=hZ_6UfegiSzkwcD8shUMtojmIxXxfjkAas_BCnS1rp4&e=> can you validate this one functionally ? You can use portainer/portainer:pr2122 to try it.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_portainer_portainer_pull_2122-23issuecomment-2D410704174&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=0fx0h4vB56iTLpw2McH1ZD6TqG_QGpbggVOB-PfMJpM&m=uHSnTkysep872hMF0-ElTCSUQZTqyC2TVr-PsxPC83g&s=kEjf_uiqwTZg5DaKwFhvCKB3e1hvqcsfTAeCcvLq-Q0&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AWGrlWZDOO3jPCna1D-5FbbXT3P9xg4oVPks5uOEGxgaJpZM4Vs6WL&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=0fx0h4vB56iTLpw2McH1ZD6TqG_QGpbggVOB-PfMJpM&m=uHSnTkysep872hMF0-ElTCSUQZTqyC2TVr-PsxPC83g&s=9bWbvwp3XOkJCjUYE87iDM7hcQeCzlUZnrgNBaThZYg&e=>.

xAt0mZ commented:
@ncresswell forgot to save parent network card, fixed it !
Also added the `--attachable` option on network creation, as we don't have it.

@deviantony updated

deviantony commented:
@ncresswell `portainer/portainer:pr2122` updated

deviantony commented on app/docker/views/networks/create/createNetworkController.js:
```diff
@@ -1,138 +1,195 @@
+
+        PluginService.networkPlugins(apiVersion < 1.25)
+          .then(function success(data) {
+            if ($scope.applicationState.endpoint.mode.provider !== 'DOCKER_SWARM_MODE')
```
Please use the form `if (condition) {}`

deviantony commented on app/docker/views/networks/create/createNetworkController.js:
```diff
@@ -1,138 +1,195 @@
+          }
+        } else {
+          var nodeName = $scope.formValues.NodeName;
+          HttpRequestHelper.setPortainerAgentTargetHeader(nodeName);
```
I believe that the following code block can be located outside of the `if/else` block:

```javascript
HttpRequestHelper.setPortainerAgentTargetHeader(nodeName);
createNetwork(networkConfiguration, userDetails, accessControlData, true);
```

Nodename should be computed in each block of the if/else as well as the last parameter of the `createNetwork` function.

deviantony commented on app/docker/views/networks/create/createNetworkController.js:
```diff
@@ -1,138 +1,195 @@
+          })
+          .then(function success() {
+            Notifications.success('Network successfully created');
+            if (reload)
```
I'm not sure to understand when we need to reload and when not?

xAt0mZ replied:
We can deploy the macvlan config network to many nodes, but we can't create it on all nodes at the same time, we should not redirect the app to `/networks` until the last network config is sent to the API.
So we reload only after the last call of config network creation, but we always reload on all cases

deviantony commented on app/docker/components/network-macvlan-form/networkMacvlanFormController.js:
```diff
@@ -0,0 +1,53 @@
+          });
+
+        var apiVersion = ctrl.applicationState.endpoint.apiVersion;
+        NetworkService.networks(
```
Retrieving the networks should be located inside the `$q.all` call.

deviantony commented on app/docker/components/network-macvlan-form/networkMacvlanForm.html:
```diff
@@ -0,0 +1,97 @@
+<div>
+  <!-- selector -->
```
I think that we should add another section title here:

```html
          <div class="col-sm-12 form-section-title">
            Macvlan configuration
          </div>
```

deviantony commented on app/docker/components/network-macvlan-form/networkMacvlanForm.html:
```diff
@@ -0,0 +1,97 @@
+  <!-- selector -->
+  <div class="col-sm-12">
+    <div class="form-group" style="margin-bottom: 0">
+      <div class="boxselector_wrapper">
```
If there is no availableNetworks, I think that we should not show the "I want to configure a network before deploying it".

UX wise, it does not make sense to show a box selector if there is only choice available. Instead, a small information message should be displayed.

xAt0mZ replied:
Hiding the box selectors in one context and displaying them in another isn't UX friendly, as the view isn't the same depending on the same "above" context (means, you can't remove things from the screen, for a same visible context, based on things not shown on the screen).

Maybe we should keep the 2 selectors visible but disable the "Creation" one if the user doesn't have config networks (turning it grey + small information message) ?

deviantony replied:
Agree, I like the view you posted.

deviantony commented on app/docker/views/networks/create/createNetworkController.js:
```diff
@@ -1,138 +1,207 @@
+          accessControlData: accessControlData,
+          reload: true
+        };
+        // configure macvlan specs
```
Please remove this comment.

deviantony commented on app/docker/views/networks/create/createNetworkController.js:
```diff
@@ -1,138 +1,207 @@
+          }
+        }
+
+        // create network
```
Please remove this comment as well.

deviantony commented:
@ncresswell implementation ready for final tests via `portainer/portainer:pr2122`.

deviantony commented:
@xAt0mZ please hide the "enable manual attachment" and "restrict external access" switches when creating a config.

xAt0mZ commented:
@deviantony updated
